File: python/0206.py
from data_strcutres import ListNode

'''
head: 1 -> 2 -> 3 -> N
curr: N

head: 2 -> 3
curr: 1 -> N

head: 3
curr: 2 -> 1 -> N

head: N
curr: 3 -> 2 -> 1 -> N
'''


class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        curr = None
        while head:
            next = head.next
            head.next = curr
            curr = head
            head = next
        return curr

File: cpp/0206.cpp
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head || !head->next) {
            return head;
        }
        ListNode* dummy = new ListNode(0, head);
        ListNode* cur = head;

        while (cur->next != nullptr) {
            ListNode* keep = cur->next->next;
            cur->next->next = dummy->next;
            dummy->next = cur->next;
            cur->next = keep;
        }
        return dummy->next;
    }
};

int main() {
    Solution sol;
    ListNode* root = new  ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
    ListNode* reversed = sol.reverseList(root);
    while (reversed != nullptr) {
        cout << reversed->val << " ";
        reversed = reversed->next;
    }
    return 0;
}
